staging/src/k8s.io/cli-runtime/pkg/printers/kyaml.go

```diff
@@ -0,0 +1,457 @@
+//
+// Any type that implements json.Marshaler will be marshaled to JSON and then
+// to YAML.
+type KYAMLPrinter struct {
+	printCount int64
+}
+
+// PrintObj prints the data as KYAML to the specified writer.
+func (p *KYAMLPrinter) PrintObj(obj runtime.Object, w io.Writer) error {
+	// We use reflect.Indirect here in order to obtain the actual value from a pointer.
+	// We need an actual value in order to retrieve the package path for an object.
+	// Using reflect.Indirect indiscriminately is valid here, as all runtime.Objects are supposed to be pointers.
+	if InternalObjectPreventer.IsForbidden(reflect.Indirect(reflect.ValueOf(obj)).Type().PkgPath()) {
+		return errors.New(InternalObjectPrinterErr)
+	}
+
+	// Print a comment shebang to indicate that this is a YAML document.
+	count := atomic.AddInt64(&p.printCount, 1)
+	if count == 1 {
+		fmt.Fprintln(w, "#!yaml")
+	}
+
+	// Always emit a document separator, which helps disambiguate between YAML
+	// and JSON.
+	fmt.Fprintln(w, "---")
+
+	switch obj := obj.(type) {
+	case *metav1.WatchEvent:
+		if InternalObjectPreventer.IsForbidden(reflect.Indirect(reflect.ValueOf(obj.Object.Object)).Type().PkgPath()) {
+			return errors.New(InternalObjectPrinterErr)
+		}
+	case *runtime.Unknown:
+		return p.fromJSON(obj.Raw, w)
+	}
+
+	if obj.GetObjectKind().GroupVersionKind().Empty() {
+		return fmt.Errorf("missing apiVersion or kind; try GetObjectKind().SetGroupVersionKind() if you know the type")
+	}
+
+	return p.fromAny(obj, w)
+}
+
+func (p *KYAMLPrinter) fromJSON(jsonBytes []byte, w io.Writer) error {
+	jsonObj, err := p.unmarshalJSON(jsonBytes)
+	if err != nil {
+		return err
+	}
+	return p.fromAny(jsonObj, w)
+}
+
+func (p *KYAMLPrinter) unmarshalJSON(jsonBytes []byte) (interface{}, error) {
+	// We are using yaml.Unmarshal here (instead of json.Unmarshal) because the
+	// Go JSON library doesn't try to pick the right number type (int, float,
+	// etc.) when unmarshalling to interface{}, it just picks float64
+	// universally. go-yaml does go through the effort of picking the right
+	// number type, so we can preserve number type throughout this process.
+	var jsonObj interface{}
+	if err := goyaml2.Unmarshal(jsonBytes, &jsonObj); err != nil {
+		return nil, fmt.Errorf("error unmarhsaling from JSON: %w", err)
+	}
+	return jsonObj, nil
+}
+
+func (p *KYAMLPrinter) fromAny(obj any, w io.Writer) error {
+	buf := &strings.Builder{}
+	if err := p.formatValue(reflect.ValueOf(obj), 0, flagDefault, buf); err != nil {
+		return err
+	}
+	fmt.Fprintf(w, "%s\n", buf.String())
+	return nil
+}
+
+type flagMask uint64
+
+const (
+	flagDefault   flagMask = 0
+	flagLazyQuote flagMask = 1 << iota
+)
+
+func (p *KYAMLPrinter) formatValue(v reflect.Value, indent int, flags flagMask, buf *strings.Builder) error {
+	if !v.IsValid() {
+		buf.WriteString("null")
+		return nil
+	}
+
+	// Handle very special things.
+	if v.CanInterface() {
+		if t, ok := v.Interface().(metav1.Time); ok {
```

Comment:
Could you add a comment here to explain why a special case is used for `metav1.Time` vs. treating it like other `json.Marshaler` implementations?

Reply:
```
JSON:
    "2024-12-11T00:10:11Z"
YAML:
    "2024-12-11T00:10:11Z"
KYAML:
    "2024-12-11T00:10:11Z"
Without this hunk:
    "2024-12-11T00:10:11Z"
```

So I guess this was a cut-paste artifact.  So much of this area is uncommented.

Reply:
Without this, some tests fail.  `v1.Time` gets hit with a 0 value, which self-unmarshals to `null`.  I guess I can catch that and just convert it to any old pointer type, to render as "null".  Am I missing something?